### PR TITLE
@W-23953663 fix(mrt): improve Brotli streaming performance

### PR DESCRIPTION
**Pull request by clavery**
## Summary

- use Brotli quality 6 by default for dynamic streaming responses instead of Node's implicit quality 11
- flush Brotli after each 32 KiB of uncompressed input so progressive SSR chunks reach the client promptly
- preserve an explicitly configured Brotli quality and all other caller-supplied options
- add regression coverage for the default quality and pre-`end()` compressed output

## Why

Node's implicit Brotli quality 11 is optimized for offline compression. On a production-like 1.6 MB Storefront Next PLP response, it consumed seconds of Lambda time and buffered useful compressed output until the response completed. This also affects traffic through an upstream CDN because the CDN requests Brotli/Gzip from MRT before passing through or transforming the response for the visitor.

Controlled warmed A/B measurements using the same application bundle showed:

| Measurement | Original | Patched | Improvement |
| --- | ---: | ---: | ---: |
| MRT-origin Brotli total | 3.659 s | 0.709 s | **5.2× faster** |
| MRT-origin post-header tail | 3.074 s | 0.077 s | **39.9× shorter** |
| CDN Brotli total | 3.810 s | 0.775 s | **4.9× faster** |

The compressed response grew from approximately 102 KB to 125 KB (about 23%) in exchange for eliminating roughly three seconds of response time and restoring progressive delivery. A local compression-only benchmark of the same 1.6 MB document improved from 1,641 ms at quality 11 to 9.45 ms at quality 6.

## Validation

- `pnpm --filter @salesforce/mrt-utilities test:agent` — 559 passing
- `pnpm --filter @salesforce/mrt-utilities typecheck:agent`
- `pnpm --filter @salesforce/mrt-utilities lint:agent`
- `pnpm --filter @salesforce/mrt-utilities format:check`
- `pnpm --filter @salesforce/mrt-utilities build`

## Companion change

Storefront Next currently bundles a separate copy of this streaming adapter. The companion draft PR applies and tests the same fix there: commerce-emu/storefront-next#2648.

### CODE REVIEW

File: packages/mrt-utilities/src/streaming/create-lambda-adapter.ts [hunk 1]
```diff
@@ -28,6 +28,80 @@ import {ServerlessRequest} from '@h4ad/serverless-adapter';
...
+ * Resolves the Brotli compression quality from the MRT_BROTLI_COMPRESSION_QUALITY
+ * environment variable, falling back to DEFAULT_BROTLI_QUALITY. Values that are not
+ * integers within the valid Brotli quality range (0-11) are ignored.
+ *
+ * @returns The Brotli quality level to use.
+ */
+function resolveBrotliQuality(): number {
+  const raw = process.env.MRT_BROTLI_COMPRESSION_QUALITY;
+  if (raw === undefined || raw.trim() === '') {
+    console.log('Using default Brotli quality:', DEFAULT_BROTLI_QUALITY);
```
Comment (kieran-sf) [inline packages/mrt-utilities/src/streaming/create-lambda-adapter.ts]: Logs just for testing, will remove

**Review by npeternel-sf — APPROVED** [review]

**Review by bendvc — CHANGES_REQUESTED** [review]
## Review: REQUEST CHANGES

### Intent Summary

The PR fixes a real production performance regression — Node.js's Brotli quality 11 was consuming 3.6+ s of Lambda CPU on large SSR responses. Lowering the default to quality 6 and adding periodic `BROTLI_OPERATION_FLUSH` calls every 32 KiB is the right approach. The implementation logic is sound; the issues below are in test coverage and a few correctness edge cases.

---

### Acceptance Criteria Scorecard

| Criterion | Status | Notes |
|---|---|---|
| Use Brotli quality 6 by default | ✅ | Implemented and tested |
| Flush every 32 KiB of uncompressed input | ✅ | Implemented and tested |
| Preserve caller-supplied Brotli quality | ⚠️ **UNTESTED** | Logic is correct but this specific path has zero test coverage — the PR's own stated criterion |
| Add regression coverage for quality + pre-`end()` output | ✅ | Two dedicated tests added |

---

### Blockers

**1. Test stub not restored in `afterEach` — real test isolation bug.**

The "default quality" test calls `createBrotliStub.restore()` as the last statement in the test body. If any assertion before that line throws, `zlib.createBrotliCompress` stays permanently stubbed for all subsequent tests — silently producing false results. The `environment variable overrides` block correctly uses `afterEach(() => sinon.restore())`; the outer tests do not. This needs to move to `afterEach`.

**2. No test for caller-supplied quality taking precedence (stated AC, zero coverage).**

The guard `if (brotliOptions?.params?.[qualityParameter] !== undefined)` is the entire mechanism fulfilling the "preserve explicit quality" acceptance criterion. That branch is never exercised. If the guard were accidentally inverted or removed, no test would catch it.

---

### Correctness Issues

**3. `MRT_BROTLI_CHUNKING_ENABLED=0` silently enables chunking.**

```typescript
const raw = process.env.MRT_BROTLI_CHUNKING_ENABLED;
if (!raw) {
  return true;
}
return raw.toLowerCase() !== 'false';
```

`'0'` is truthy, so `'0' !== 'false'` → `true` (chunking **on**). Unix convention says `0` = false. The other two resolver functions use `raw.trim() === ''` for the empty check, making this one inconsistent. Either handle `'0'` as falsy or document that only `'false'` disables it.

**4. No warning on invalid env var values.**

`MRT_BROTLI_COMPRESSION_QUALITY=foo` silently falls back to 6. On a Lambda where someone sets the var and nothing changes, there's no trace of why. A single `console.warn` in the invalid branch is all this needs.

**5. `flush()` called unconditionally under potential back-pressure.**

`flush()` is called after a `write()` that may have returned `false` (stream at high-water mark). Calling `flush()` during a drain cycle can produce unexpected chunk interleaving. Low-probability in Lambda's ephemeral stream context, but worth guarding with `if (accepted)`.

---

### Additional Test Gaps

- `MRT_BROTLI_COMPRESSION_QUALITY=0` (minimum valid value — exercises the `>= BROTLI_MIN_QUALITY` branch specifically)
- Non-quality `brotliOptions.params` are preserved through the spread
- `MRT_BROTLI_CHUNKING_ENABLED=0` documents/pins the current behavior

**Fragile timing oracles:** Two tests use `await new Promise(resolve => setTimeout(resolve, 50))` as "no data has arrived yet." On a loaded CI runner this is flaky. A `setImmediate` or `await Promise.resolve()` chain would be deterministic.

---

### Docs

The "Tuning Brotli compression" section is accurate. One gap: it says "Set to `false` to disable" chunking but doesn't mention that `0` is silently ignored.

---

### Nit

Resolver functions are called per-invocation inside `createExpressResponse` / `createCompressionStream`. Since Lambda env vars are static across invocations, caching these at module level would avoid re-parsing on every request. Not a bug, just a smell.

---

Overall the fix is correct and the performance improvement is real. The two blockers above are straightforward to address — the isolation bug especially should be fixed before this merges.

**Review by bendvc — CHANGES_REQUESTED** [review]
## Review: REQUEST CHANGES

### Intent Summary

The PR fixes a real production performance regression — Node.js's Brotli quality 11 was consuming 3.6+ s of Lambda CPU on large SSR responses. Lowering the default to quality 6 and adding periodic `BROTLI_OPERATION_FLUSH` calls every 32 KiB is the right approach. The implementation logic is sound; the issues below are in test coverage and a few correctness edge cases.

---

### Acceptance Criteria Scorecard

| Criterion | Status | Notes |
|---|---|---|
| Use Brotli quality 6 by default | ✅ | Implemented and tested |
| Flush every 32 KiB of uncompressed input | ✅ | Implemented and tested |
| Preserve caller-supplied Brotli quality | ⚠️ **UNTESTED** | Logic is correct but this specific path has zero test coverage — the PR's own stated criterion |
| Add regression coverage for quality + pre-`end()` output | ✅ | Two dedicated tests added |

---

### Blockers

**1. Test stub not restored in `afterEach` — real test isolation bug.**

The "default quality" test calls `createBrotliStub.restore()` as the last statement in the test body. If any assertion before that line throws, `zlib.createBrotliCompress` stays permanently stubbed for all subsequent tests — silently producing false results. The `environment variable overrides` block correctly uses `afterEach(() => sinon.restore())`; the outer tests do not. This needs to move to `afterEach`.

**2. No test for caller-supplied quality taking precedence (stated AC, zero coverage).**

The guard `if (brotliOptions?.params?.[qualityParameter] !== undefined)` is the entire mechanism fulfilling the "preserve explicit quality" acceptance criterion. That branch is never exercised. If the guard were accidentally inverted or removed, no test would catch it.

---

### Correctness Issues

**3. `MRT_BROTLI_CHUNKING_ENABLED=0` silently enables chunking.**

```typescript
const raw = process.env.MRT_BROTLI_CHUNKING_ENABLED;
if (!raw) {
  return true;
}
return raw.toLowerCase() !== 'false';
```

`'0'` is truthy, so `'0' !== 'false'` → `true` (chunking **on**). Unix convention says `0` = false. The other two resolver functions use `raw.trim() === ''` for the empty check, making this one inconsistent. Either handle `'0'` as falsy or document that only `'false'` disables it.

**4. No warning on invalid env var values.**

`MRT_BROTLI_COMPRESSION_QUALITY=foo` silently falls back to 6. On a Lambda where someone sets the var and nothing changes, there's no trace of why. A single `console.warn` in the invalid branch is all this needs.

**5. `flush()` called unconditionally under potential back-pressure.**

`flush()` is called after a `write()` that may have returned `false` (stream at high-water mark). Calling `flush()` during a drain cycle can produce unexpected chunk interleaving. Low-probability in Lambda's ephemeral stream context, but worth guarding with `if (accepted)`.

---

### Additional Test Gaps

- `MRT_BROTLI_COMPRESSION_QUALITY=0` (minimum valid value — exercises the `>= BROTLI_MIN_QUALITY` branch specifically)
- Non-quality `brotliOptions.params` are preserved through the spread
- `MRT_BROTLI_CHUNKING_ENABLED=0` documents/pins the current behavior

**Fragile timing oracles:** Two tests use `await new Promise(resolve => setTimeout(resolve, 50))` as "no data has arrived yet." On a loaded CI runner this is flaky. A `setImmediate` or `await Promise.resolve()` chain would be deterministic.

---

### Docs

The "Tuning Brotli compression" section is accurate. One gap: it says "Set to `false` to disable" chunking but doesn't mention that `0` is silently ignored.

---

### Nit

Resolver functions are called per-invocation inside `createExpressResponse` / `createCompressionStream`. Since Lambda env vars are static across invocations, caching these at module level would avoid re-parsing on every request. Not a bug, just a smell.

---

Overall the fix is correct and the performance improvement is real. The two blockers above are straightforward to address — the isolation bug especially should be fixed before this merges.
